### ISSUE TITLE
GuassianFitter returns different results for same data

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#152]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/152) by Dan Becker (Bitbucket: [danbek](https://bitbucket.org/danbek), GitHub: [danbek](https://github.com/danbek)).**

----------------------------------------

Download the attached file and then run this code multiple times. You will find that the fit parameters are different each time. The differences are small (certainly less than the uncertainty in the parameters) but I still find this disturbing. Aren't computers deterministic? Is this a "feature" of this particular data, or does this always happen and I never noticed it before?

```
#!python

import mass
import numpy as np

d = np.load('fitter.npz')
n = d['n']
bins = d['bins']
    
fitter = mass.GaussianFitter()
(p,pcov) = fitter.fit(n, bins); print('yyy', p[0], p[1])
```

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Okay, this is a weird one.

Yes, the computation is deterministic, *mostly*. The exception occurs when you try to fit a function where the starting guess at a parameter lands on the lower or upper limit of the allowed range. Here, the parameter that does that is the constant background level. It has to be non-negative, but with no data "in the wings" of your sample data set, the initial guess at the parameter is indeed zero.

So if you try the above computation as 

```
(p,pcov) = fitter.fit(n+0.1, bins); print('yyy', p[0], p[1])
(p,pcov) = fitter.fit(n+0.1, bins); print('yyy', p[0], p[1])
(p,pcov) = fitter.fit(n+0.1, bins); print('yyy', p[0], p[1])
```

...you'll see that you get the same answer each time.

I think the solution here is to make the initial guess as to the Gaussian's constant background never be zero.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Make fits deterministic, even when apparent background level=0.

Mass.GaussianFitter was giving random answers for certain test cases, if
the starting-point estimate of the background level was exactly 0. We
make that estimate be no less than 0.5/N, for N data points. Fixes #152.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Make fits deterministic, even when apparent background level=0.

Mass.GaussianFitter was giving random answers for certain test cases, if
the starting-point estimate of the background level was exactly 0. We
make that estimate be no less than 0.5/N, for N data points. Fixes #152.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Removing milestone: v0.7 (automated comment)
